saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/Saml2AuthenticationToken.java

```diff
@@ -38,6 +38,32 @@
 
 	private final String saml2Response;
 
+	private AbstractSaml2AuthenticationRequest authenticationRequest;
```

Comment:
Since this won't change, I think it can be `final`.

Reply:
If I do this in the new constructor:
```java
public Saml2AuthenticationToken(RelyingPartyRegistration relyingPartyRegistration, String saml2Response,
			AbstractSaml2AuthenticationRequest authenticationRequest) {
	this(relyingPartyRegistration, saml2Response);
	this.authenticationRequest = authenticationRequest;
}
```
I can't make it `final` since it could be initialized inside the first constructor.

Reply:
I think the constructors should be inverted. For example, do:

```java
public Saml2AuthenticationToken(RelyingPartyRegistration relyingPartyRegistration, String saml2Response) {
    this(relyingPartyRegistration, saml2Response, null);
```

Reply:
Nice!